File: packages/smart-reply/ios/Plugin/SmartReply.swift
import Foundation
import MLKitSmartReply

@objc public class SmartReply: NSObject {
    private static let statusSuccess = "SUCCESS"
    private static let statusNotSupportedLanguage = "NOT_SUPPORTED_LANGUAGE"
    private static let statusNoReply = "NO_REPLY"

    @objc public func suggestReplies(_ options: SuggestRepliesOptions, completion: @escaping (_ result: SuggestRepliesResult?, _ error: Error?) -> Void) {
        MLKitSmartReply.SmartReply.smartReply().suggestReplies(for: options.messages) { result, error in
            if let error = error {
                completion(nil, error)
                return
            }
            completion(self.createResult(result), nil)
        }
    }

    private func createResult(_ result: SmartReplySuggestionResult?) -> SuggestRepliesResult {
        guard let result = result else {
            return SuggestRepliesResult(status: SmartReply.statusNoReply, suggestions: [])
        }
        var suggestions: [String] = []
        if result.status == .success {
            suggestions = result.suggestions.map { $0.text }
        }
        return SuggestRepliesResult(status: mapStatus(result.status), suggestions: suggestions)
    }

    private func mapStatus(_ status: SmartReplyResultStatus) -> String {
        switch status {
        case .notSupportedLanguage:
            return SmartReply.statusNotSupportedLanguage
        case .noReply:
            return SmartReply.statusNoReply
        default:
            return SmartReply.statusSuccess
        }
    }
}

File: packages/smart-reply/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/smartreply/SmartReply.java
package io.capawesome.capacitorjs.plugins.mlkit.smartreply;

import androidx.annotation.NonNull;
import com.google.mlkit.nl.smartreply.SmartReplyGenerator;
import com.google.mlkit.nl.smartreply.SmartReplySuggestion;
import com.google.mlkit.nl.smartreply.SmartReplySuggestionResult;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.classes.options.SuggestRepliesOptions;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.classes.results.SuggestRepliesResult;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.interfaces.NonEmptyResultCallback;
import java.util.ArrayList;
import java.util.List;

public class SmartReply {

    private static final String STATUS_SUCCESS = "SUCCESS";
    private static final String STATUS_NOT_SUPPORTED_LANGUAGE = "NOT_SUPPORTED_LANGUAGE";
    private static final String STATUS_NO_REPLY = "NO_REPLY";

    @NonNull
    private final SmartReplyGenerator smartReplyGenerator;

    public SmartReply() {
        this.smartReplyGenerator = com.google.mlkit.nl.smartreply.SmartReply.getClient();
    }

    public void close() {
        smartReplyGenerator.close();
    }

    public void suggestReplies(@NonNull SuggestRepliesOptions options, @NonNull NonEmptyResultCallback<SuggestRepliesResult> callback) {
        smartReplyGenerator
            .suggestReplies(options.getMessages())
            .addOnSuccessListener(result -> callback.success(createResult(result)))
            .addOnFailureListener(callback::error);
    }

    @NonNull
    private SuggestRepliesResult createResult(@NonNull SmartReplySuggestionResult result) {
        List<String> suggestions = new ArrayList<>();
        if (result.getStatus() == SmartReplySuggestionResult.STATUS_SUCCESS) {
            for (SmartReplySuggestion suggestion : result.getSuggestions()) {
                suggestions.add(suggestion.getText());
            }
        }
        return new SuggestRepliesResult(mapStatus(result.getStatus()), suggestions);
    }

    @NonNull
    private String mapStatus(int status) {
        switch (status) {
            case SmartReplySuggestionResult.STATUS_NOT_SUPPORTED_LANGUAGE:
                return STATUS_NOT_SUPPORTED_LANGUAGE;
            case SmartReplySuggestionResult.STATUS_NO_REPLY:
                return STATUS_NO_REPLY;
            default:
                return STATUS_SUCCESS;
        }
    }
}
